src/modules/ekf2/EKF/mag_control.cpp

```diff
@@ -41,228 +41,249 @@
-				_state.mag_I = mag_earth_pred;
+			const Dcmf R_to_body = quatToInverseRotMat(_state.quat_nominal);
+			_state.mag_B = _mag_lpf.getState() - (R_to_body * mag_earth_pred);
 
-				const Dcmf R_to_body = quatToInverseRotMat(_state.quat_nominal);
-				_state.mag_B = _mag_lpf.getState() - (R_to_body * mag_earth_pred);
+		} else {
+			// Use the last magnetometer measurements to reset the field states
+			// calculate initial earth magnetic field states
+			_state.mag_I = _R_to_earth * _mag_lpf.getState();
+			_state.mag_B.zero();
+		}
 
-			} else {
-				// Use the last magnetometer measurements to reset the field states
-				// calculate initial earth magnetic field states
-				_state.mag_I = _R_to_earth * _mag_lpf.getState();
-				_state.mag_B.zero();
-			}
+		ECL_DEBUG("resetting mag I: [%.3f, %.3f, %.3f], B: [%.3f, %.3f, %.3f]",
+			  (double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2),
+			  (double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
+			 );
 
-			ECL_DEBUG("resetting mag I: [%.3f, %.3f, %.3f], B: [%.3f, %.3f, %.3f]",
-				(double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2),
-				(double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
-				);
+		resetMagCov();
 
-			resetMagCov();
-		}
+		has_realigned_yaw = true;
 
-		if (!has_realigned_yaw && canResetMagHeading()) {
-			has_realigned_yaw = resetMagHeading();
-		}
+	} else if (!magFieldStrengthDisturbed(mag - _state.mag_B)) {
+		has_realigned_yaw = resetMagHeading(mag - _state.mag_B);
+	}
 
-		if (has_realigned_yaw) {
-			_mag_yaw_reset_req = false;
-			_control_status.flags.yaw_align = true;
+	if (has_realigned_yaw) {
+		_control_status.flags.yaw_align = true;
+
+		if (_control_status.flags.in_air) {
 			_control_status.flags.mag_aligned_in_flight = true;
 
 			// record the time for the magnetic field alignment event
 			_flt_mag_align_start_time = _imu_sample_delayed.time_us;
+		}
 
-			// Handle the special case where we have not been constraining yaw drift or learning yaw bias due
-			// to assumed invalid mag field associated with indoor operation with a downwards looking flow sensor.
-			if (_mag_inhibit_yaw_reset_req) {
-				_mag_inhibit_yaw_reset_req = false;
-				// Zero the yaw bias covariance and set the variance to the initial alignment uncertainty
-				P.uncorrelateCovarianceSetVariance<1>(12, sq(_params.switch_on_gyro_bias * _dt_ekf_avg));
-			}
+		// Handle the special case where we have not been constraining yaw drift or learning yaw bias due
+		// to assumed invalid mag field associated with indoor operation with a downwards looking flow sensor.
+		if (isTimedOut(_aid_src_mag_heading.time_last_fuse, (uint32_t)5e6)
+		    && isTimedOut(_aid_src_mag.time_last_fuse, (uint32_t)5e6)
+		    && isTimedOut(_aid_src_gnss_yaw.time_last_fuse, (uint32_t)5e6)
+		    && isTimedOut(_aid_src_ev_yaw.time_last_fuse, (uint32_t)5e6)
```

Comment:
Maybe also consider NE aiding (e.g.: GNSS vel and pos) as those are making the yaw angle observable in presence of body accelerations (i.e.: when `_yaw_angle_observable` is true).

Reply:
Done.